### PR TITLE
Speed up DecisionTree training by partitioning the order arrays

### PR DESCRIPTION
**Pull request by twotabbies**
This should help somewhat with https://github.com/haifengl/smile/issues/257 and https://github.com/haifengl/smile/issues/70

### CODE REVIEW

**Comment by haifengl** [comment]
Thanks!

File: core/src/main/java/smile/classification/DecisionTree.java [hunk 1]
```diff
@@ -744,13 +759,13 @@ private double impurity(int[] count, int n) {
 
         switch (rule) {
             case GINI:
-                impurity = 1.0;
+                double squared_sum = 0;
```
Comment (myui) [inline core/src/main/java/smile/classification/DecisionTree.java]: @haifengl Did you review this modification?

File: core/src/main/java/smile/classification/DecisionTree.java [hunk 2]
```diff
@@ -982,6 +997,119 @@ public DecisionTree(Attribute[] attributes, double[][] x, int[] y, int maxNodes,
...
+
+    /**
+     * Modifies an array in-place by partitioning the range from low (inclusive) to high
+     * (exclusive) so that all elements i for which left[i - low] is true come before all elements
+     * for which it is false, but element ordering is otherwise preserved. The number of true
+     * elements in left must equal split-low. buffer is scratch space large enough to hold all
+     * elements for which left is false.
+     */
+    private void partitionArray(int[] a, int low, int split, int high, IntPredicate goesLeft, int[] buffer) {
+        int n = high - low;
```
Comment (myui) [inline core/src/main/java/smile/classification/DecisionTree.java]: `n` is not used.

File: core/src/main/java/smile/classification/DecisionTree.java [hunk 3]
```diff
@@ -648,58 +671,47 @@ public boolean split(PriorityQueue<TrainNode> nextSplits) {
                 throw new IllegalStateException("Split a node with invalid feature.");
             }
 
-            int n = x.length;
+            int n = high - low;
```
Comment (myui) [inline core/src/main/java/smile/classification/DecisionTree.java]: `n` is not used.

File: core/src/main/java/smile/classification/DecisionTree.java [hunk 4]
```diff
@@ -982,6 +997,119 @@ public DecisionTree(Attribute[] attributes, double[][] x, int[] y, int maxNodes,
...
+        int k = 0;
+        for (int i = low; i < high; i++) {
+            if (goesLeft.test(a[i])) {
+                a[j++] = a[i];
+            } else {
+                buffer[k++] = a[i];
+            }
+        }
+        if (k != high - split || j != split) {
+            System.err.println("Messed up partition: " + low + ".." + split + ".." + high + " ended up splitting at " + j);
```
Comment (myui) [inline core/src/main/java/smile/classification/DecisionTree.java]: Better to throw IllegalStateException for fail-fast if this condition should not be met.

**Comment by myui** [comment]
@haifengl @twotabbies 

I'm not yet fully understand this approach but is [partitionOrder](https://github.com/haifengl/smile/pull/448/files#diff-aa1c5dce2c959f46efdcb90067706154R725) considers recursion in [split](https://github.com/haifengl/smile/pull/448/files#diff-aa1c5dce2c959f46efdcb90067706154R732)?

In [trueChild.split(null)](https://github.com/haifengl/smile/pull/448/files#diff-aa1c5dce2c959f46efdcb90067706154R732), the destructive `partitionOrder` method modified `order` field. Then, [falseChild.findBestSplit()](https://github.com/haifengl/smile/pull/448/files#diff-aa1c5dce2c959f46efdcb90067706154R737) uses modified `order` using `split/high`.

`split(null)` (depth-first split) is not used now as discussed in https://github.com/haifengl/smile/pull/359 though.

File: core/src/main/java/smile/classification/DecisionTree.java (hunk 1 above)
Comment (twotabbies) [reply core/src/main/java/smile/classification/DecisionTree.java]: This is supposed to be equivalent - both old and new compute $1 - \sum_i c_i^2 / n^2$, but the old approach has a division per class, while the new has only a single division.

File: core/src/main/java/smile/classification/DecisionTree.java (hunk 2 above)
Comment (twotabbies) [reply core/src/main/java/smile/classification/DecisionTree.java]: Thanks!

File: core/src/main/java/smile/classification/DecisionTree.java (hunk 3 above)
Comment (twotabbies) [reply core/src/main/java/smile/classification/DecisionTree.java]: Thanks!

File: core/src/main/java/smile/classification/DecisionTree.java (hunk 4 above)
Comment (twotabbies) [reply core/src/main/java/smile/classification/DecisionTree.java]: Sigh. There's always one left-over bit of debugging that should have been removed. The unit tests do pass.

**Comment by twotabbies** [comment]
> @haifengl @twotabbies
> 
> I'm not yet fully understand this approach but is [partitionOrder](https://github.com/haifengl/smile/pull/448/files#diff-aa1c5dce2c959f46efdcb90067706154R725) considers recursion in [split](https://github.com/haifengl/smile/pull/448/files#diff-aa1c5dce2c959f46efdcb90067706154R732)?
> 
> In [trueChild.split(null)](https://github.com/haifengl/smile/pull/448/files#diff-aa1c5dce2c959f46efdcb90067706154R732), the destructive `partitionOrder` method modifies `order` field. Then, [falseChild.findBestSplit()](https://github.com/haifengl/smile/pull/448/files#diff-aa1c5dce2c959f46efdcb90067706154R737) uses modified `order` using `split/high`.
> 
> `split(null)` (depth-first split) is not used now as discussed in #359 though.

This is fine. partitionOrder() has partitioned the order arrays so that the true child owns low...split-1, and the false child owns split...high-1. A node and (importantly) all its descendants can modify only the range that it owns, so whether the true node is descended immediately (depth-first) or later (when pulled from nextSplits) it won't affect the expansion of the false node.

**Comment by twotabbies** [comment]
Thanks for the code comments. Since this has been merged already, I'll put together another pull request with the modifications.

**Comment by myui** [comment]
hmm, `(6)` uses `order` modified at `(3)` but range of modification is safe for (6).
```
-- level1
(1) partitionOrder(low1, split1, high1, ..)
(2) trueChild1.findBestSplit() -> findBestSplit(.., low1, split1, ..)
  -- level2
  (3) partitionOrder(low1, split2, split1, ..)
  (4) trueChild2.findBestSplit() -> findBestSplit(.., low1, split2, ..)
  (5) falseChild1.findBestSplit() -> findBestSplit(.., split2, high1, ..)
(6) falseChild1.findBestSplit() -> findBestSplit(.., split1, high1, ..)
```

Suppose best-first split as follows 

```
      (1)
     /   \
   (2)   (2)'
  /  \   /  \
(3) (3)' (4) (4)
    /  \
  (5) (5)

1. root splits
2. left side of 2nd level splits
3. right side of 3rd level (3) splits
4. right side of 2nd level splits 
```

level3 and level2 uses different low/high range but never overlaps, and thus destructive modifications to `order/originalOrder` field are safe (?).

```
-- level3 2nd leftmost node 
(1) node3.partitionOrder()
(2) trueChild3.findBestSplit() 
(3) falseChild3.findBestSplit() 
-- level2 rightmost node
(4) node2.partitionOrder() -- uses order/originalOrder modified in (1)
(5) trueChild2.findBestSplit()
(6) falseChild2.findBestSplit()
```

File: core/src/main/java/smile/classification/DecisionTree.java (hunk 1 above)
Comment (myui) [reply core/src/main/java/smile/classification/DecisionTree.java]: hmm, right. I have little concern about overflow at `squared_sum += ` though.

**Pull request merged**